flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1513,13 +1517,20 @@ private CompletableFuture<Void> waitForTerminatingJob(
                                                     throwable));
                                 });
 
-        return jobManagerTerminationFuture.thenAcceptAsync(
+        // keep track of the job as outstanding, if not done
+        if (!jobManagerTerminationFuture.isDone()) {
+            submittedAndWaitingTerminationJobIDs.add(jobId);
+        }
+
+        return FutureUtils.thenAcceptAsyncIfNotDone(
```

Comment:
AFAICT this change along, without any of the other stuff, would already solve the issue in question.
If no JM exists for this job then this runs synchronously in the main thread.

Reply:
I agree, managing the duplicate job IDs explicitly is merely a helper to managing the `jobManagerRunnerTerminationFutures`, but it did seem to me that it closes a gap in the job submission flow

Reply:
I think we need to keep the ID to prevent the issue in mentioned https://github.com/apache/flink/pull/21849/files#r1212786448.